ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerImpl.java

```diff
@@ -247,23 +281,80 @@ public SafeFuture<ExecutionPayloadHeader> builderGetHeader(
         executionPayloadContext.getParentHash());
 
     return executionBuilderClient
-        .get()
+        .orElseThrow()
         .getHeader(slot, Bytes48.ZERO, executionPayloadContext.getParentHash())
         .thenApply(ExecutionLayerManagerImpl::unwrapResponseOrThrow)
         .thenApply(
-            builderBidV1SignedMessage ->
-                getExecutionHeaderFromBuilderBid(builderBidV1SignedMessage, slot))
+            builderBidV1SignedMessage -> getHeaderFromBuilderBid(builderBidV1SignedMessage, slot))
         .thenPeek(
-            executionPayloadHeader ->
-                LOG.trace(
-                    "builderGetHeader(slot={}, pubKey={}, parentHash={}) -> {}",
-                    slot,
-                    Bytes48.ZERO,
-                    executionPayloadContext.getParentHash(),
-                    executionPayloadHeader));
+            executionPayloadHeader -> {
+              // store that we haven't fallen back for this slot
+              slotToLocalElFallbackData.put(slot, Optional.empty());
+              LOG.trace(
+                  "builderGetHeader(slot={}, pubKey={}, parentHash={}) -> {}",
+                  slot,
+                  Bytes48.ZERO,
+                  executionPayloadContext.getParentHash(),
+                  executionPayloadHeader);
+            })
+        .exceptionallyCompose(
+            error -> {
+              LOG.error(
+                  "builderGetHeader returned an error. Falling back to local execution engine",
+                  error);
+              return doFallbackToLocal(localExecutionPayload, slot);
+            });
+  }
+
+  @Override
+  public SafeFuture<ExecutionPayload> builderGetPayload(
+      final SignedBeaconBlock signedBlindedBeaconBlock) {
+
+    checkArgument(
+        signedBlindedBeaconBlock.getMessage().getBody().isBlinded(),
+        "SignedBeaconBlock must be blind");
+
+    final UInt64 slot = signedBlindedBeaconBlock.getSlot();
+
+    final Optional<Optional<SafeFuture<ExecutionPayload>>> maybeFallbackPayload =
+        Optional.ofNullable(slotToLocalElFallbackData.get(slot));
```

Comment:
The `Optional<Optional<..>>` makes this a bit confusing, particularly with the `.get().get()` later. I can see why it's needed and is probably better than an Optional var that is null.

It may be clearer if we can "unwrap" each level to a local var as we go though so we don't have `.get()` calls through the code.

Reply:
The reason why I added a new optional is because intellij complains null-testing an Optional.
I'll unwrap it

Reply:
Yeah I figured.  To be clear, by "unwrap" I don't mean going back to null but just do essentially:

```
Optional<Optional<Foo>> start = l...
if (start.isEmpty()) {
  return bah;
}
Optional<Foo> middle = start.get();
if (middle.isEmpty()) {
  return bah;
}
Foo foo = middle.get();
```

Just so we don't wind up with .get().get() and .get().isEmpty() and other hard to decipher stuff.

Reply:
yes yes, I got it. See the latest commit :)